config/src/test/scala/com/typesafe/config/impl/ConfigSubstitutionTest.scala

```diff
@@ -924,7 +924,7 @@ class ConfigSubstitutionTest extends TestUtils {
 
     @Test
     def substSelfReferenceIndirect() {
-        val obj = parseObject("""a=1, b=${a}, a=${b}""")
+        val obj = parseObject("""b=${a}, a=${b}""")
```

Comment:
The details are spelled out here:
 https://github.com/typesafehub/config/blob/master/HOCON.md#duplicate-keys-and-object-merging
and here:
 https://github.com/typesafehub/config/blob/master/HOCON.md#self-referential-substitutions

Merging of duplicates is essential to the library; it's what's happening when you do:
 ```
  foo.x = 10
  foo.y = 12
```
In this case the `foo` field is duplicated (as `foo : { x : 10 }` and `foo : { y : 12 }`) and then merged. Merging is also used for "inheritance" like this: https://github.com/typesafehub/config/blob/master/README.md#inheritance
And it's used for plain old overrides, and to combine multiple config files into one. This is the `withFallback` method in the API.

It is always possible the test is wrong, though offhand it looks ok to me. But we really have to know why Java 8 changes the outcome, because it's almost certainly a bug that Java 8 changes the outcome, even if Java 8 gets it right and Java 6 got it wrong, we need to fix the undefined/unreliable behavior that resulted in it being wrong in one of them.